// metaapi.cloud-copyfactory-sdk/dist/copyFactory.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _httpClient = require('./clients/httpClient');

var _httpClient2 = _interopRequireDefault(_httpClient);

var _configuration = require('./clients/copyFactory/configuration.client');

var _configuration2 = _interopRequireDefault(_configuration);

var _history = require('./clients/copyFactory/history.client');

var _history2 = _interopRequireDefault(_history);

var _trading = require('./clients/copyFactory/trading.client');

var _trading2 = _interopRequireDefault(_trading);

var _domain = require('./clients/domain.client');

var _domain2 = _interopRequireDefault(_domain);

var _logger = require('./logger');

var _logger2 = _interopRequireDefault(_logger);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * MetaApi CopyFactory copy trading API SDK
 */
class CopyFactory {

  /**
   * Enables using Log4js logger with extended log levels for debugging instead of
   * console.* functions. Note that log4js configuration performed by the user.
   */
  static enableLog4jsLogging() {
    _logger2.default.useLog4js();
  }

  /**
   * Constructs CopyFactory class instance
   * @param {String} token authorization token
   * @param {Object} opts connection options
   * @param {String} [opts.domain] domain to connect to
   * @param {String} [opts.extendedTimeout] timeout for extended http requests in seconds
   * @param {Number} [opts.requestTimeout] timeout for http requests in seconds
   */
  constructor(token, opts = {}) {
    this._domain = opts.domain || 'agiliumtrade.agiliumtrade.ai';
    let requestTimeout = opts.requestTimeout || 10;
    let requestExtendedTimeout = opts.extendedTimeout || 70;
    const retryOpts = opts.retryOpts || {};
    this._httpClient = new _httpClient2.default(requestTimeout, requestExtendedTimeout, retryOpts);
    this._domainClient = new _domain2.default(this._httpClient, token, this._domain);
    this._configurationClient = new _configuration2.default(this._domainClient);
    this._historyClient = new _history2.default(this._domainClient);
    this._tradingClient = new _trading2.default(this._domainClient);
  }

  /**
   * Returns CopyFactory configuration API
   * @returns {ConfigurationClient} configuration API
   */
  get configurationApi() {
    return this._configurationClient;
  }

  /**
   * Returns CopyFactory history API
   * @return {HistoryClient} history API
   */
  get historyApi() {
    return this._historyClient;
  }

  /**
   * Returns CopyFactory trading API
   * @return {TradingClient} trading API
   */
  get tradingApi() {
    return this._tradingClient;
  }
}
exports.default = CopyFactory;
