// metaapi.cloud-copyfactory-sdk/dist/clients/copyFactory/streaming/stopoutListenerManager.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _metaApi = require('../../metaApi.client');

var _metaApi2 = _interopRequireDefault(_metaApi);

var _randomstring = require('randomstring');

var _randomstring2 = _interopRequireDefault(_randomstring);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Stopout event listener manager
 */
class StopoutListenerManager extends _metaApi2.default {

  /**
   * Constructs stopout listener manager instance
   * @param {DomainClient} domainClient domain client
   */
  constructor(domainClient) {
    super(domainClient);
    this._domainClient = domainClient;
    this._stopoutListeners = {};
    this._errorThrottleTime = 1000;
  }

  /**
   * Returns the dictionary of stopout listeners
   * @returns {Object} dictionary of stopout listeners
   */
  get stopoutListeners() {
    return this._stopoutListeners;
  }

  /**
   * Adds a stopout listener
   * @param {StopoutListener} listener 
   * @param {String} [accountId] account id
   * @param {String} [strategyId] strategy id
   * @param {Number} [sequenceNumber] event sequence number
   * @returns {String} stopout listener id
   */
  addStopoutListener(listener, accountId, strategyId, sequenceNumber) {
    const listenerId = _randomstring2.default.generate(10);
    this._stopoutListeners[listenerId] = listener;
    this._startStopoutEventJob(listenerId, listener, accountId, strategyId, sequenceNumber);
    return listenerId;
  }

  /**
   * Removes stopout listener by id
   * @param {String} listenerId listener id 
   */
  removeStopoutListener(listenerId) {
    delete this._stopoutListeners[listenerId];
  }

  async _startStopoutEventJob(listenerId, listener, accountId, strategyId, sequenceNumber) {
    let throttleTime = this._errorThrottleTime;

    while (this._stopoutListeners[listenerId]) {
      const opts = {
        url: '/users/current/stopouts/stream',
        method: 'GET',
        qs: {
          previousSequenceNumber: sequenceNumber,
          subscriberId: accountId,
          strategyId: strategyId,
          limit: 1000
        },
        headers: {
          'auth-token': this._token
        },
        json: true
      };
      try {
        const packets = await this._domainClient.requestCopyFactory(opts, true);
        await listener.onStopout(packets);
        throttleTime = this._errorThrottleTime;
        if (this._stopoutListeners[listenerId] && packets.length) {
          sequenceNumber = packets.slice(-1)[0].sequenceNumber;
        }
      } catch (error) {
        await new _promise2.default(res => setTimeout(res, throttleTime));
        throttleTime = Math.min(throttleTime * 2, 30000);
      }
    }
  }

}
exports.default = StopoutListenerManager;
