// metaapi.cloud-sdk/dist/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DrawdownListener = exports.RiskManagement = exports.MetaStats = exports.TransactionListener = exports.UserLogListener = exports.StopoutListener = exports.CopyFactory = exports.MemoryHistoryStorage = exports.SynchronizationListener = exports.HistoryStorage = undefined;

var _metaApi = require('./metaApi/metaApi');

var _metaApi2 = _interopRequireDefault(_metaApi);

var _historyStorage = require('./metaApi/historyStorage');

var _historyStorage2 = _interopRequireDefault(_historyStorage);

var _memoryHistoryStorage = require('./metaApi/memoryHistoryStorage');

var _memoryHistoryStorage2 = _interopRequireDefault(_memoryHistoryStorage);

var _synchronizationListener = require('./clients/metaApi/synchronizationListener');

var _synchronizationListener2 = _interopRequireDefault(_synchronizationListener);

var _metaapi = require('metaapi.cloud-copyfactory-sdk');

var _metaapi2 = _interopRequireDefault(_metaapi);

var _metaapi3 = require('metaapi.cloud-metastats-sdk');

var _metaapi4 = _interopRequireDefault(_metaapi3);

var _metaapi5 = require('metaapi.cloud-risk-management-sdk');

var _metaapi6 = _interopRequireDefault(_metaapi5);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = _metaApi2.default;
exports.HistoryStorage = _historyStorage2.default;
exports.SynchronizationListener = _synchronizationListener2.default;
exports.MemoryHistoryStorage = _memoryHistoryStorage2.default;
exports.CopyFactory = _metaapi2.default;
exports.StopoutListener = _metaapi.StopoutListener;
exports.UserLogListener = _metaapi.UserLogListener;
exports.TransactionListener = _metaapi.TransactionListener;
exports.MetaStats = _metaapi4.default;
exports.RiskManagement = _metaapi6.default;
exports.DrawdownListener = _metaapi5.DrawdownListener;
